database/sqlite/sqlite_health.c

```diff
@@ -0,0 +1,1347 @@
+            (unsigned int)sqlite3_column_int64(res, 6),
+            sqlite3_column_type(res, 25) == SQLITE_NULL ? "-" : format_value_and_unit(new_value_string, 100, (calculated_number) sqlite3_column_double(res, 25), (char *) sqlite3_column_text(res, 19), -1),
+            sqlite3_column_type(res, 26) == SQLITE_NULL ? "-" : format_value_and_unit(old_value_string, 100, (calculated_number) sqlite3_column_double(res, 26), (char *) sqlite3_column_text(res, 19), -1),
+            (long unsigned int)sqlite3_column_int64(res, 27),
+            (sqlite3_column_int64(res, 10) & HEALTH_ENTRY_FLAG_SILENCED)?"true":"false");
+
+        buffer_strcat(wb, "\t\t\"info\":\"");
+        buffer_strcat(wb, (char *) sqlite3_column_text(res, 20));
+        buffer_strcat(wb, "\",\n");
+
+        if(unlikely(sqlite3_column_int64(res, 10) & HEALTH_ENTRY_FLAG_NO_CLEAR_NOTIFICATION)) {
+            buffer_strcat(wb, "\t\t\"no_clear_notification\": true,\n");
+        }
+
+        buffer_strcat(wb, "\t\t\"value\":");
+        if (sqlite3_column_type(res, 25) == SQLITE_NULL)
+            buffer_strcat(wb, "null");
+        else
+            buffer_rrd_value(wb, (calculated_number) sqlite3_column_double(res, 25));
+        buffer_strcat(wb, ",\n");
+
+        buffer_strcat(wb, "\t\t\"old_value\":");
+        if (sqlite3_column_type(res, 26) == SQLITE_NULL)
+            buffer_strcat(wb, "null");
+        else
+            buffer_rrd_value(wb, (calculated_number) sqlite3_column_double(res, 26));
+        buffer_strcat(wb, "\n");
+
+        buffer_strcat(wb, "\t}");
+
+        freez(edit_command);
+    }
+
+    failed:
+    rc = sqlite3_finalize(res);
+    if (unlikely(rc != SQLITE_OK))
+        error_report("Failed to finalize statement for SQL_SELECT_HEALTH_LOG");
+}
+
+#define SQL_SELECT_ALL_HEALTH_LOG(guid,limit) "SELECT unique_id, alarm_id FROM health_log_%s order by unique_id desc LIMIT %u;", guid,limit
+void sql_health_alarm_log_select_all(BUFFER *wb, RRDHOST *host) {
+    sqlite3_stmt *res = NULL;
+    int rc;
+    char command[MAX_HEALTH_SQL_SIZE + 1];
+
+    if (unlikely(!db_meta)) {
+        if (default_rrd_memory_mode == RRD_MEMORY_MODE_DBENGINE)
+            error_report("Database has not been initialized");
+        return;
+    }
+
+    char uuid_str[GUID_LEN + 1];
+    uuid_unparse_lower(host->host_uuid, uuid_str);
+    uuid_str[8] = '_';
+    uuid_str[13] = '_';
+    uuid_str[18] = '_';
+    uuid_str[23] = '_';
+
+    snprintfz(command, MAX_HEALTH_SQL_SIZE, SQL_SELECT_ALL_HEALTH_LOG(uuid_str, host->health_log.max));
+
+    rc = sqlite3_prepare_v2(db_meta, command, -1, &res, 0);
+    if (unlikely(rc != SQLITE_OK)) {
+        error_report("Failed to prepare statement for SQL_SELECT_ALL_HEALTH_LOG");
+        return;
+    }
+
+    int count=0;
+    while (sqlite3_step(res) == SQLITE_ROW) {
+        if (count) buffer_strcat(wb, ",");
+        health_alarm_entry_sql2json(wb, (uint32_t) sqlite3_column_int64(res, 0), (uint32_t) sqlite3_column_int64(res, 1), host);
```

Comment:
Its probably better to prepare the statement before calling the function and pass it as a parameter and reset the parameters as needed

Reply:
Actually, removed both `health_alarm_entry_sql2json` and `sql_health_alarm_log_select_all`, where not used. If required in later point will be re-added.